omegaconf/basecontainer.py

```diff
@@ -747,6 +751,23 @@ def prepand(full_key: str, parent_type: Any, cur_type: Any, key: Any) -> str:
 
         return full_key
 
+    # If item has a Container annotation like Dict[str, int] it validates against the target element_type
+    # if it doesn't have an annotation it returns the raw value
+    def _resolve_container_assignment(self, item: Any) -> Any:
```

Comment:
This function is doing too much, which also leads to it having an opaque name.
It's doing some assignment validation (and raising an exception), and unwrappingg the object in some cases.

Assignment validation is done in validate_set. It's not clear to me why you want to perform a partial validation from various places.
Unwrapping can happen as a part of assignment. (_set_value() on DictConfig and ListConfig).

Reply:
I'm going to tackle this between today and tomorrow.